processing/src/main/java/io/druid/segment/IndexMerger.java

```diff
@@ -821,6 +829,14 @@ public void close() throws IOException
                                       ? null
                                       : Ints.asList(dims[i]);
           forwardDimWriters.get(i).write(listToWrite);
+          if (listToWrite == null || listToWrite.isEmpty()) {
+            // empty row; add to the nullRows bitmap
+            nullRowsList.get(i).add(rowCount);
+          } else if (dimHasNullFlags.get(i) && listToWrite.size() == 1 && listToWrite.get(0) == 0) {
+            // If this dimension has the null/empty str in its dictionary, a row with a single-valued dimension
+            // that matches the null/empty str's dictionary ID should also be added to nullRowsList.
+            nullRowsList.get(i).add(rowCount);
+          }
```

Comment:
add(..) on MutableBitmap is expensive and leads to a performance regression here because this will, in many cases, get called for num_rows * num_dimensions